pkg/activator/throttler.go

```diff
@@ -0,0 +1,173 @@
+	breakers := make(map[RevisionID]queue.Breaker)
+	return &Throttler{breakers: breakers, breakerParams: breakerParams, revisionInformer: revisionInformer, logger: logger}
+}
+
+// Throttler allows to update the max concurrency of registered Breakers
+type Throttler struct {
+	breakers         map[RevisionID]queue.Breaker
+	revisionInformer v1alpha1.RevisionInformer
+	breakerParams    queue.BreakerParams
+	logger           *zap.SugaredLogger
+	mux              sync.RWMutex
+}
+
+// Add potentially creates a new breaker for the revision
+// returns true if a new entry was added
+func (t *Throttler) AddIfAbsent(rev RevisionID) bool {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	if _, ok := t.breakers[rev]; !ok {
+		breaker := queue.NewBreaker(t.breakerParams)
+		t.breakers[rev] = *breaker
+		return true
+	}
+	return false
+}
+
+// Remove deletes the breaker from the bookkeeping
+func (t *Throttler) Remove(rev RevisionID) {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	delete(t.breakers, rev)
+}
+
+// UpdateCapacity update the max concurrency of the Breaker corresponding to the revision
+func (t *Throttler) UpdateCapacity(rev RevisionID, size int32) error {
+	t.mux.RLock()
+	defer t.mux.RUnlock()
+	var containerConcurrency int32
+	revision, exists, err := t.getRevision(rev.Namespace, rev.Name, t.revisionInformer)
+	if !exists {
+		return fmt.Errorf("no revision was found for: %s", rev)
+	}
+	if err != nil {
+		return err
+	}
+	containerConcurrency = int32(revision.Spec.ContainerConcurrency)
+
+	if containerConcurrency == 0 {
+		containerConcurrency = t.breakerParams.MaxConcurrency
+	}
+	if breaker, ok := t.breakers[rev]; ok {
+		err := breaker.UpdateConcurrency(size * containerConcurrency)
+		if err != nil {
+			return err
+		}
+	} else {
+		return fmt.Errorf("no breaker was found for the received revision: %s", rev)
+	}
+	return nil
+}
+
+// Try executes the `function` on the Breaker
+// Returns true if the function was executed, false otherwise
+// The function is not executed either if the Breaker doesn't exist or it is full
+func (t *Throttler) Try(rev RevisionID, function func()) error {
+	t.mux.RLock()
+	if breaker, ok := t.breakers[rev]; ok {
+		t.mux.RUnlock()
+		ok := breaker.Maybe(function)
+		if !ok {
+			return errors.New("activator throttler is in overflow")
+		}
+		return nil
+	}
+	t.mux.RUnlock()
+	return fmt.Errorf("breaker is not found for the revision %s", rev)
+}
+
+func (t *Throttler) getRevision(namespace, name string, informer v1alpha1.RevisionInformer) (*v1alpha12.Revision, bool, error) {
+	var revision *v1alpha12.Revision
+	revisionKey := cache.ExplicitKey(namespace + "/" + name)
+	rev, exists, err := informer.Informer().GetIndexer().Get(revisionKey)
+	if exists && rev != nil {
+		revision = rev.(*v1alpha12.Revision)
+	}
+	return revision, exists, err
+}
+
+// Helper defines methods handler methods for endpoints observer
+type Helper struct {
+}
+
+// UpdateEndpoints communicates the new number of endpoints to the Throttler
```

Comment:
Please update the comment to reflect reality.

Reply:
Done